### ISSUE TITLE
Usage for a Union

### ISSUE THREAD

**Issue by borissmidt** [isocpp/CppCoreGuidelines#573]
the usage of a union is undefined i.e. ???

I have programmed in C for 6 months on very limited embedded system (8bit 4k memory) and i used unions regularly.
I used unions for two reasons:
-The reuse of static memory for a shared I/O buffer this was to avoid a lot of rewriting. i.e.
```
struct buffer{
   uint8_t rx[128];
   uint8_t tx[128];
}
```
can become a union in the code without changing too much about the code if the io buffer is halve duplex

-Type punning to deliberately violate the type system. To avoid a copy. Otherwise this required undefined behaviour like casting a pointer to a different type.
So i have added it to the guide but its only useful for limited systems and should only be used there,
(it might be a better idea to say its only useful in embedded C/C++)

I already made a pull request with examples, but i forgot to open an issue for this.

**Comment by MikeGitb** [comment]
As this is an area, I'm personally interested in, I have a question regarding the second use case:
> Type punning to deliberately violate the type system.

Have you verified that you actually can circumvent the type system in c++ in that way? If I recall correctly, the rules for unions differ in c and c++ and I thought it is UB in c++ to write to one member and read from the other if they are of different type. I have to admit though, that I'm not too familiar with EC++ and I believe that many "normal" c++ compilers guarantee defined behavior just like in c code.

**Comment by borissmidt** [comment]
Im reading the standard now for the union. 
it states that "The size of a union is sufficient to contain the largest
of its non-static data members" and later in the paragraph " All non-static data members of a union object have the same address."

`
#include <stdint.h>
#include <iostream>
union Collor{
   uint8_t ARGB[4];
   uint32_t color;
};

union CollorStruct{
   struct{
     uint8_t B; // endianess
     uint8_t G;
     uint8_t R;
     uint8_t A;
   };
   uint32_t color;
};

union TestFloat{
  float X;
  uint32_t data;
};

union TestFun{
  //usize_t ptr;
  bool (*bas)();
  bool (*foo)(int);
  bool (*bar)(int,int);
};

bool Bas(){
  return true;
}

bool Foo(int x){
  std::cout<< "X is " << x << '\n';
  return true;
}

bool Bar(int y, int z){
    std::cout<<"Y is " << y << " Z is " << z << '\n';

  return false;
}

int main(int argc, char const *argv[]) {
  //type punning example
  union Collor col;
  col.color = 0x0070FF75;
  std::cout<< "address of collor and ARGB " << &col.color <<
  '\n'<<  &col.ARGB  << '\n';

  std::cout<< std::hex << col.color <<
  '\n'<<  (int) col.ARGB[3] << (int)  col.ARGB[2]
  << (int) col.ARGB[1] <<  (int) col.ARGB[0] << '\n' ;

  col.ARGB[1] = 01;
  std::cout<< std::hex << col.color <<
  '\n'<<  (int) col.ARGB[3] << (int)  col.ARGB[2]
  << (int) col.ARGB[1] <<  (int) col.ARGB[0] << '\n' ;

  CollorStruct colS;
  colS.color = 0x77CC56AB;
  std::cout<< std::hex << colS.color <<
  '\n'<<  (int) colS.A << (int)  colS.R
  << (int) colS.G <<  (int) colS.B << '\n';
  colS.G = 0x50;
  std::cout<< std::hex << colS.color <<
  '\n'<<  (int) colS.A << (int)  colS.R
  << (int) colS.G <<  (int) colS.B << '\n';

  union TestFloat f;
  f.X = 100.0f;
  std::cout<<  f.X << '\n';
  f.data = f.data | 0x80000000; // flip sign to make - obsolete
  std::cout<< f.X << '\n';

  //test functions
  TestFun fun;
  fun.foo = Foo;
  std::cout<< fun.foo(1) << '\n';
  fun.bar = Bar;
  std::cout<< fun.bar(5,10) << '\n';
  //clang allows for 'vardiac functions'
  fun.foo = Foo;
  fun.bas();
  fun.bar(10,11);
//clang++ and g++ allows for 'vardiac functions' it seems
// which should really be undefined
  fun.bar = Bar;
  fun.bas();
  fun.foo(50);
  fun.bar(13,18);
  return 0;
}
`
i guess i should get a medal for writing  obsecure C++ 

it also seems c++ allows for non unions with non trivial data members 
but you have to delete is manually 
u.m. ~ M();
new (&u.n) N;

So there is no real definition what happens when you acces different union members, but it is stated that they share the same memory adress. so you will just overwrite the other types binary data while any read is to the same element. a bit like a void* but extended so it is always alligned.

**Comment by MikeGitb** [comment]
I can't search the standard right now, but the non-normative text [on cppreference.com](http://en.cppreference.com/w/cpp/language/union) states that 
> [...], and it's **undefined behavior** to read from the member of the union that wasn't most recently written.

**Comment by borissmidt** [comment]
Yes the standard paper doesn't define what should happen when you read out a member which was not most recently set. But it also said that every variable in the container will have the same adress. So i suspect that this forces the compilers to align the data. So when you read the data you will jsut read the binary data at that address. I guess i should try this example on the Microsoft compiler as well.

**Comment by MikeGitb** [comment]
Irrespective of what the standard says, I believe that most compilers explicitly allow type punning through unions, so I don't have a problem with this per se. I just wanted to say that if it is not supported by the standard, I'm not sure if it should be in the core guidelines. But lets see, what the maintainers have to say.

**Comment by borissmidt** [comment]
I guess they should define it in c++17 :)

**Comment by MikeGitb** [comment]
That would certainly be nice. 

But even then, things like 
```
union TestFloat{
    float X;
    uint32_t data;
};
```
would open aliasing issues, if we'd pass around references / pointers to individual elements.

**Comment by borissmidt** [comment]
Wouldn't it give the same issues as if you share a pointer to the same array or object twice to a function? But with the extra problem that you cannot assume that unint8_t * doesn't aliase with unint32_t*

**Comment by MikeGitb** [comment]
Well, in that particular case it wouldn't be an additional burden, as `uint8_t*` is often a typedef for  `unsigned char*`, which is already allowed to point to an `uint32_t`. However, in other cases (such as `float*` and `uint32_t*`) it might noticeably hamper the compiler's optimization capabilities if such usage of unions would be allowed without some additional restrictions.

**Comment by borissmidt** [comment]
Yea,  i have had a job in embedded C on an 8 bit cpu so a uint8_t wouldn't be the same as a uint32_t. But yes the compiler will then have to assume that two different typed pointers can aliase. But i guess this is more of a problem for the programmer than for the compiler. 
So this would result in the same problem as if someone static_casted a struct{ float x} * to a struct {uint32_t y}* and pass both to a function. 
So a union could be naively seen as a struct{ largest data type} * const with implicit casting.
